source/docs/software/basic-programming/code-formatting.rst

```diff
@@ -0,0 +1,209 @@
+
+Spotless can also be used as a :doc:`CI check <robot-code-ci>`. The check is ran with ``./gradlew spotlessCheck``.
+
+Explanation of Options
+^^^^^^^^^^^^^^^^^^^^^^
+
+Each ``format`` section highlights formatting of custom files in the project. The ``java`` and ``groovyGradle`` are natively supported by spotless, so they are defined differently.
+
+Breaking this down, we can split this into multiple parts.
+
+- Formatting Java
+- Formatting Gradle files
+- Formatting XML files
+- Formatting Miscellaneous files
+
+They are all similar, except for some small differences that will be explained. The below example will highlight the ``java {}`` block.
+
+.. code-block:: groovy
+
+   java {
+      target fileTree('.') {
+         include '**/*.java'
+         exclude '**/build/**', '**/build-*/**'
+      }
+      toggleOffOn()
+      googleJavaFormat()
+      removeUnusedImports()
+      trimTrailingWhitespace()
+      endWithNewline()
+   }
+
+Let's explain what each of the options mean.
+
+.. code-block:: groovy
+
+   target fileTree('.') {
+      include '**/*.java'
+      exclude '**/build/**', '**/build-*/**'
+   }
+
+The above example tells spotless where our Java classes are and to exclude the ``build`` directory. The rest of the options are fairly self-explanatory.
+
+- ``toggleOffOn()`` adds the ability to have spotless ignore specific portions of a project. The usage looks like the following
+
+.. code-block:: java
+
+   // format:off
+
+   public void myWeirdFunction() {
+
+   }
+
+   // format:on
+
+- ``googleJavaFormat()`` tells spotless to format according to the `Google Style Guide <https://google.github.io/styleguide/javaguide.html>`__
+- ``removeUnusedImports()`` will remove any unused imports from any of your java classes
+- ``trimTrailingWhitespace()`` will remove any extra whitespace at the end of your lines
+- ``endWithNewline()`` will add a newline character to the end of your classes
+
+In the ``groovyGradle`` block, there is a ``greclipse`` option. This is the formatter that spotless uses to format gradle files.
+
+Additionally, there is a ``eclipseWtp`` option in the ``xml`` block. This stands for "Gradle Web Tools Platform" and is the formatter to format ``xml`` files. Teams not using any XML files may wish to not include this configuration.
+
+.. note:: A full list of configurations is available on the `Spotless README <https://github.com/diffplug/spotless>`__
+
+wpiformat
+---------
+
+Requirements
+^^^^^^^^^^^^
+
+- `Python 3.6 or higher <https://www.python.org/>`__
+- clang-format (included with `LLVM <https://releases.llvm.org/download.html>`__)
+
+.. important:: Windows is not currently supported at this time! Installing LLVM with Clang **will** break normal robot builds if installed on Windows.
+
+You can install `wpiformat <https://github.com/wpilibsuite/styleguide/blob/main/wpiformat/README.rst>`__ by typing ``pip3 install wpiformat``.
```

Comment:
Should we add where they need to type this?

Reply:
Done